apps/src/templates/CodeWorkspace.jsx

```diff
@@ -1,71 +1,188 @@
+  },
+  headerButtonIcon: {
+    lineHeight: '24px',
+    paddingRight: 8,
+    fontSize: 15
+  },
   noPadding: {
     padding: 0
+  },
+  bold: {
+    fontWeight: 'bold'
+  },
+  chevron: {
+    fontSize: 18,
+  },
+  runningChevron: {
+    color: color.dark_charcoal
   }
 };
 
-var CodeWorkspace = function (props) {
-  return (
-    <span id="codeWorkspaceWrapper">
-      <ProtectedStatefulDiv id="headers" dir={props.localeDirection}>
-        <div id="codeModeHeaders">
-          <div id="toolbox-header" className="workspace-header">
-            <i id="hide-toolbox-icon" style={commonStyles.hidden} className="fa fa-chevron-circle-right"/>
-            <span>{props.editCode ? msg.toolboxHeaderDroplet() : msg.toolboxHeader()}</span>
-          </div>
-          <div id="show-toolbox-header" className="workspace-header" style={commonStyles.hidden}>
-            <i id="show-toolbox-icon" className="fa fa-chevron-circle-right"></i>
-            <span>{msg.showToolbox()}</span>
-          </div>
-          <div id="show-code-header" className="workspace-header workspace-header-button">
-            <span>
-              <img src="/blockly/media/applab/blocks_glyph.gif" className="blocks-glyph" />
-              <i className="fa fa-code"/>
-              <span style={styles.noPadding}>{msg.showCodeHeader()}</span>
-            </span>
-          </div>
-          {!props.readonlyWorkspace && <div id="clear-puzzle-header" className="workspace-header workspace-header-button">
-            <span>
-              <i className="fa fa-undo"/>
-              <span style={styles.noPadding}>{msg.clearPuzzle()}</span>
-            </span>
-          </div>}
-          <div id="versions-header" className="workspace-header workspace-header-button">
-            <span>
-              <i className="fa fa-clock-o"/>
-              <span style={styles.noPadding}>{msg.showVersionsHeader()}</span>
-            </span>
-          </div>
-          <div id="workspace-header" className="workspace-header">
-            <span id="workspace-header-span">
-              {props.readonlyWorkspace ? msg.readonlyWorkspaceHeader() : msg.workspaceHeaderShort()}
-            </span>
-            <div id="blockCounter">
-              <div id="blockUsed" className='block-counter-default'>
+var CodeWorkspace = React.createClass({
+  propTypes: {
+    localeDirection: React.PropTypes.oneOf(['rtl', 'ltr']).isRequired,
+    editCode: React.PropTypes.bool.isRequired,
+    readonlyWorkspace: React.PropTypes.bool.isRequired,
+    showDebugger: React.PropTypes.bool,
+    isRunning: React.PropTypes.bool
+  },
+
+  shouldComponentUpdate: function (nextProps) {
+    // This component is current semi-protected. We don't want to completely
+    // disallow rerendering, since that would prevent us from being able to
+    // update styles. However, we do want to prevent property changes that would
+    // change the DOM structure.
+    Object.keys(nextProps).forEach(function (key) {
```

Comment:
This is a bit convoluted :stuck_out_tongue_winking_eye: but clever as an interim solution.  Hopefully the few controls in the headers will be managed by React soon and we can just protect inside the content areas.